### PR TITLE
💡 Node `assert`❔

### PR DESCRIPTION
**Pull request by smnandre**
Acting as a _spinoff_ to [#4165 Add tag to declare variable types](https://github.com/twigphp/Twig/issues/4165)
(as I don't want to disrupt the main conversation too much 😅)

The idea is to introduce assertions.

I'm opening this to gather feedback and ideas. I'd like to see if there's something interesting here—or not. :)

This could enhance static analysis, define contracts in templates, and serve as a validator during includes/rendering.

```twig
{% assert name string %}
{% assert size in ['sm', 'md', 'lg'] %}
```

Open to any  feedback :)

### CODE REVIEW

**Comment by smnandre** [comment]
## Usage / syntax examples 


| Langage                                                                                  | Code example                                                 |
|------------------------------------------------------------------------------------------|-----------------------------------------------------------------|
| [**PHP**](https://www.php.net/manual/en/function.assert.php)                             | `assert(in_array($size, ['sm', 'md']));`                        |
| [**Python**](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) | `assert size in ['sm', 'md']`                                   |
| [**Java**](https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html)  | `assert java.util.Arrays.asList("sm", "md").contains(size);`    |
| [**C**](https://en.cppreference.com/w/c/error/assert)                                    | `assert(strcmp(size, "sm") == 0 \|\| strcmp(size, "md") == 0);` |
| [**JavaScript**](https://developer.mozilla.org/en-US/docs/Web/API/console/assert)        | `console.assert(["sm", "md"].includes(size));`                  |
| [**C++**](https://en.cppreference.com/w/cpp/error/assert)                                | `assert(size == "sm" \|\| size == "md");`                       |
| [**Ruby**](https://ruby-doc.org/core-3.1.2/Kernel.html#method-i-raise)                   | `raise 'Assertion failed' unless ['sm', 'md'].include?(size)`   |


(Table made with some gpt -  for representational purposes only)

File: src/Node/AssertNode.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        $compiler
+            ->write(' if (!')
+            ->subcompile($this->getNode('expr'))
+            ->raw(") {\n")
+            ->indent()
+            ->write("throw new RuntimeError('Failed');\n")
+            ->outdent()
+            ->raw("}\n")
+        ;
```
Comment (Kocal) [inline src/Node/AssertNode.php:44]: What about using `assert()` function?  
With OPcache configuration, you can literally remove `assert()` calls from the code... or here you planned to not write anything if production?

Comment (smnandre) [reply src/Node/AssertNode.php:44]: That's an idea too !  I would say a node would be easier to parse / use by node visitors but i may be wrong here 😅 

> or here you planned to not write anything if production?

I planned... _nothing_ 🙃   the "idea" just emerged earlier tonight during the "var/type" discussion

Comment (stof) [reply src/Node/AssertNode.php:44]: The suggestion is about the compiled code, not about the Twig API

Actually, to be able to use PHP's `assert` function which is a no-op at runtime in production mode, it needs to be a Twig tag and not Twig function usable in an expression.

**Comment by drjayvee** [comment]
When it comes to asserting variable types, that can be left over to extensions like [TwigQI](https://github.com/alisqi/TwigQI), once `types` is merged.

I do like the idea of using `assert` for expressions.

There's currently no way to intentionally throw an Exception of trigger a warning (aside from something silly like `{{ [].foo }}`).

I'd propose that the `assert` tag should have an optional message as part of its spec:
```
{% assert <expression> [message string] %}
```

@ruudk is working on compiling Twig templates to PHP in such a way that PHPStan can work its magic. If the `assert` tag gets a similar treatment, with properly typed variables, we can take a huge leap in template quality.

**Comment by fabpot** [comment]
> @ruudk is working on compiling Twig templates to PHP in such a way that PHPStan can work its magic. If the `assert` tag gets a similar treatment, with properly typed variables, we can take a huge leap in template quality.

I'm curious to see how that would look like... but also if that's something we can include in the "default" Twig compilation (I suppose this is based on attributes/annotations with no runtime side-effects).

**Comment by ruudk** [comment]
You can read more about TwigStan here: https://github.com/twigphp/Twig/discussions/4233

File: src/Node/AssertNode.php (hunk 1 above)
Comment (smnandre) [reply src/Node/AssertNode.php:44]: 👍

**Comment by fabpot** [comment]
Now that we have the `types` tag, can we close this one?

**Comment by smnandre** [comment]
> Now that we have the `types` tag, can we close this one?


In the end, i think this "assert" would be more related to assertions / tests than strict typing... 

But I do agree we should focus on the "types" tag and see  how / why it is used, the features it enables, ... 

(and we can reopen here if needed later)

**Pull request closed without merging**